## **Importing Libraries**

In [41]:
import numpy as np
import pandas as pd

## **Inspecting Data**

In [63]:
df_true=pd.read_csv("/content/True.csv",error_bad_lines=False)

<ipython-input-63-0ea9cff5223e>:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  df_true=pd.read_csv("/content/True.csv",error_bad_lines=False)


In [75]:
df_fake=pd.read_csv("/content/Fake (2).csv")

In [70]:
df_true

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
19564  Philippine Congress agrees to restore rights c...   
19565  Trump's threat to 'destroy' North Korea is wro...   
19566  U.S. urges Congolese security forces' restrain...   
19567  Exclusive: From Russia with fuel - North Korea...   
19568                                           Merkel d   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
19564  MANILA (Reuters) - The Philippines  Congress l...     worldnews   
19565  BERLIN (Reuters) - Germany and the United Stat...     worldnews   
19566  WASHINGTON (Reuters) - The U.S. State Departme...     worldnews   
19567  MOSCOW (Reuters) - At least eight North Korean...     worldnews   
19568                                                NaN           NaN   

                      date  
0       December 31, 2017   
1       December 29, 2017   
2       December 31, 2017   
3       December 30, 2017   
4       December 29, 2017   
...                    ...  
19564  September 20, 2017   
19565  September 20, 2017   
19566  September 20, 2017   
19567  September 20, 2017   
19568                  NaN  

[19569 rows x 4 columns]

In [76]:
df_fake

title  \
0       Donald Trump Sends Out Embarrassing New Year’...   
1       Drunk Bragging Trump Staffer Started Russian ...   
2       Sheriff David Clarke Becomes An Internet Joke...   
3       Trump Is So Obsessed He Even Has Obama’s Name...   
4       Pope Francis Just Called Out Donald Trump Dur...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text      subject  \
0      Donald Trump just couldn t wish all Americans ...         News   
1      House Intelligence Committee Chairman Devin Nu...         News   
2      On Friday, it was revealed that former Milwauk...         News   
3      On Christmas day, Donald Trump announced that ...         News   
4      Pope Francis used his annual Christmas Day mes...         News   
...                                                  ...          ...   
23476  21st Century Wire says As 21WIRE reported earl...  Middle-east   
23477  21st Century Wire says It s a familiar theme. ...  Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...  Middle-east   
23479  21st Century Wire says Al Jazeera America will...  Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...  Middle-east   

                    date  
0      December 31, 2017  
1      December 31, 2017  
2      December 30, 2017  
3      December 29, 2017  
4      December 25, 2017  
...                  ...  
23476   January 16, 2016  
23477   January 16, 2016  
23478   January 15, 2016  
23479   January 14, 2016  
23480   January 12, 2016  

[23481 rows x 4 columns]

In [77]:
df_true["label"]=1
df_fake["label"]=0

In [78]:
df_merge=pd.concat([df_true,df_fake],axis=0)
df_merge

title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3      WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4      SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   
...                                                  ...           ...   
23476  21st Century Wire says As 21WIRE reported earl...   Middle-east   
23477  21st Century Wire says It s a familiar theme. ...   Middle-east   
23478  Patrick Henningsen  21st Century WireRemember ...   Middle-east   
23479  21st Century Wire says Al Jazeera America will...   Middle-east   
23480  21st Century Wire says As 21WIRE predicted in ...   Middle-east   

                     date  label  
0      December 31, 2017       1  
1      December 29, 2017       1  
2      December 31, 2017       1  
3      December 30, 2017       1  
4      December 29, 2017       1  
...                   ...    ...  
23476    January 16, 2016      0  
23477    January 16, 2016      0  
23478    January 15, 2016      0  
23479    January 14, 2016      0  
23480    January 12, 2016      0  

[43050 rows x 5 columns]

In [79]:
df_merge['label'].value_counts()

0    23481
1    19569
Name: label, dtype: int64

In [80]:
df_merge.drop(['title','subject','date'],axis=1,inplace=True)

In [81]:
df_merge

text  label
0      WASHINGTON (Reuters) - The head of a conservat...      1
1      WASHINGTON (Reuters) - Transgender people will...      1
2      WASHINGTON (Reuters) - The special counsel inv...      1
3      WASHINGTON (Reuters) - Trump campaign adviser ...      1
4      SEATTLE/WASHINGTON (Reuters) - President Donal...      1
...                                                  ...    ...
23476  21st Century Wire says As 21WIRE reported earl...      0
23477  21st Century Wire says It s a familiar theme. ...      0
23478  Patrick Henningsen  21st Century WireRemember ...      0
23479  21st Century Wire says Al Jazeera America will...      0
23480  21st Century Wire says As 21WIRE predicted in ...      0

[43050 rows x 2 columns]

In [82]:
# shuffling the dataset
df_merge.sample(frac=1)

text  label
15249  BARACK AND MICHELLE   GOOD RIDDANCE!Columbia U...      0
12096  Michelle Obama told Oprah Winfrey in her final...      0
8414   WASHINGTON (Reuters) - Two senior lawmakers on...      1
12777  LONDON (Reuters) - Brexit minister David Davis...      1
7977   Donald Trump is an unapologetic Islamophobic b...      0
...                                                  ...    ...
4733   WASHINGTON (Reuters) - Iran would face tighter...      1
10752  Remember when experts came out after Hillary c...      0
12526  I wonder if they ll get the same treatment Din...      0
309    Graham-Cassidy, the latest and worst in a line...      0
2812   (Reuters) - Co-host of MSNBC’s “Morning Joe” a...      1

[43050 rows x 2 columns]

In [83]:
df_merge.isna().sum()

text     1
label    0
dtype: int64

In [84]:
df_merge.dropna(inplace=True)

In [85]:
df_merge.isna().sum()

text     0
label    0
dtype: int64

## **Text Data Cleaning**

In [86]:
import re
# Define a function to remove special characters from a text
def remove_special_characters(text):
    # Use re.sub to remove special characters

    cleaned_text = re.sub(r'[^A-Za-z0-9\s]', '', text)
    cleaned_text = cleaned_text.lower()
    return cleaned_text

# Apply the function to the text column
df_merge['text'] = df_merge['text'].apply(remove_special_characters)

# Print the modified DataFrame
print(df_merge)

                                                    text  label
0      washington reuters  the head of a conservative...      1
1      washington reuters  transgender people will be...      1
2      washington reuters  the special counsel invest...      1
3      washington reuters  trump campaign adviser geo...      1
4      seattlewashington reuters  president donald tr...      1
...                                                  ...    ...
23476  21st century wire says as 21wire reported earl...      0
23477  21st century wire says it s a familiar theme w...      0
23478  patrick henningsen  21st century wireremember ...      0
23479  21st century wire says al jazeera america will...      0
23480  21st century wire says as 21wire predicted in ...      0

[43049 rows x 2 columns]


### Saperating features and target variables

In [87]:
x=df_merge['text']
y=df_merge['label']

In [88]:
x

0        washington reuters  the head of a conservative...
1        washington reuters  transgender people will be...
2        washington reuters  the special counsel invest...
3        washington reuters  trump campaign adviser geo...
4        seattlewashington reuters  president donald tr...
                               ...                        
23476    21st century wire says as 21wire reported earl...
23477    21st century wire says it s a familiar theme w...
23478    patrick henningsen  21st century wireremember ...
23479    21st century wire says al jazeera america will...
23480    21st century wire says as 21wire predicted in ...
Name: text, Length: 43049, dtype: object

## **Splitting the dataset into train and test**

In [89]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets with a test size of 0.5 (50%)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.5)

## Text Data Vectorization using TF-IDF

In [90]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Create a TF-IDF vectorizer with specific settings
tfidf = TfidfVectorizer(stop_words='english', max_df=0.8)

# Fit and transform the training data
x_train_tfidf = tfidf.fit_transform(x_train)

# Transform the testing data using the same vectorizer
x_test_tfidf = tfidf.transform(x_test)


In [91]:
x_train_tfidf

<21524x156057 sparse matrix of type '<class 'numpy.float64'>'
	with 3320394 stored elements in Compressed Sparse Row format>

## ***Machine Learning Algorithms ***

## Logistic Regression

In [92]:
from sklearn.linear_model import LogisticRegression

In [93]:
LR=LogisticRegression()
LR.fit(x_train_tfidf,y_train)

LogisticRegression()

In [94]:
pred_lr=LR.predict(x_test_tfidf)

In [95]:
from sklearn.metrics import confusion_matrix
CM=confusion_matrix(y_test,pred_lr)

In [96]:
CM

array([[11539,   164],
       [  129,  9693]])

### Checking accuracy score

In [98]:
from sklearn.metrics import accuracy_score
acc=accuracy_score(y_test,pred_lr)
acc

0.9863879210220674

In [97]:
# Import necessary libraries
from sklearn.metrics import precision_score

# Calculate precision
precision = precision_score(y_test,pred_lr)

# Print the precision score
print("Precision:", precision)


Precision: 0.9833620777112712


### **Decision Tree classifier**

In [99]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion="entropy",random_state=42)
classifier.fit(x_train_tfidf,y_train)
# criterion parameter is the function used to measure the quality of a split of tree nodes
# and it is  "gini" or "entropy"
# entropy is more complex since it makes use of algorithms so the calculation of the Gini Index will be faster

DecisionTreeClassifier(criterion='entropy', random_state=42)

In [100]:
y_pred=classifier.predict(x_test_tfidf)

In [101]:
CM=confusion_matrix(y_test,y_pred)

In [102]:
CM

array([[11641,    62],
       [   71,  9751]])

In [103]:
from sklearn.metrics import accuracy_score
acc=accuracy_score(y_test,pred_lr)
acc

0.9863879210220674

### Random Forest classifier

In [104]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier(n_estimators=10,criterion="entropy",random_state=42)
classifier.fit(x_train_tfidf,y_train)
y_pred=classifier.predict(x_test_tfidf)
from sklearn.metrics import confusion_matrix
CM=confusion_matrix(y_test,y_pred)
CM

array([[11386,   317],
       [  642,  9180]])

In [105]:
acc_rf=accuracy_score(y_test,y_pred)

In [106]:
acc_rf

0.9554471544715447

I conducted four classification tasks, and I achieved an impressive 98% accuracy with both the logistic regression and decision tree classifiers in two of those tasks.